File: Queue-implementation/queue.rb
# Implementation of Queue in Ruby

# class Node
class Node 
 attr_accessor :data, :next
    # Initialize a new node with data and next pointer
 def initialize(data)
   @data = data
   @next = nil
 end
end

# Queue Class
class Queue 

    def initialize 
        @front = nil
        @rear = nil
    end

#    Enqueue Operation
    def enqueue(value)
       new_node = Node.new(value)
       if @rear.nil?
        @front = @rear = new_node
       else 
        @rear.next = new_node
        @rear = new_node
       end
    end

    # Dequeue Operation
    def dequeue 
        if @front.nil?
            puts "Queue is empty"
            return nil
        end
        dequeued_value = @front.data
        @front = @front.next
        @front = @rear = nil if @front.nil? # If the queue becomes empty
        dequeued_value
    end
    # Peek Operation
    def peek 
    if @front.nil?
        puts "Queue is empty"
        return nil
    end
    @front.data
    end
    # is Empty Boolean
    def is_empty?
        @front.nil?
    end

    # Print queue elements
    def print_queue
        if @front.nil?
            puts "Queue is empty"
            return 
        end

        result = "Front -> "
        current = @front

        while current
            result += "[#{current.data}] -> "
            current = current.next
        end
        result += "nil"
        puts result
    end
end

# Queue Usage 

q = Queue.new

q.enqueue(10)
q.enqueue(20)
q.enqueue(30)
q.print_queue   # Front -> [10] -> [20] -> [30] -> nil

puts "Front element: #{q.peek}" # 10
puts "Dequeued: #{q.dequeue}"   # 10
q.print_queue   # Front -> [20] -> [30] -> nil

q.enqueue(40)
q.print_queue   # Front -> [20] -> [30] -> [40] -> nil

# OutPut:
# Front -> [10] -> [20] -> [30] -> nil
# Front element: 10
# Dequeued: 10
# Front -> [20] -> [30] -> nil
# Front -> [20] -> [30] -> [40] -> nil

File: Queue-implementation/queue.cs
// Implementation of Queue data stucture in C sharp (C#)

using System;

namespace QueueImplementation
{
    // Node class
    class Node
    {
        public int Data;
        public Node? Next;

        public Node(int data)
        {
            Data = data;
            Next = null;
        }
    }

    // Queue class
    class Queue
    {
        private Node? front;
        private Node? rear;

        public Queue()
        {
            front = null;
            rear = null;
        }

        // Enqueue operation
        public void Enqueue(int value)
        {
            Node newNode = new Node(value);
            if (rear == null)
            {
                front = rear = newNode;
            }
            else
            {
                rear.Next = newNode;
                rear = newNode;
            }
        }

        // Dequeue operation
        public int? Dequeue()
        {
            if (front == null)
            {
                Console.WriteLine("Queue is empty!");
                return null;
            }

            int value = front.Data;
            front = front.Next;

            if (front == null)
            {
                rear = null;
            }

            return value;
        }

        // Peek operation
        public int? Peek()
        {
            if (front == null)
            {
                Console.WriteLine("Queue is empty!");
                return null;
            }
            return front.Data;
        }

        // Check if empty
        public bool IsEmpty()
        {
            return front == null;
        }


        // Print queue elements
        public void PrintQueue()
        {
            if (front == null)
            {
                Console.WriteLine("Queue is empty!");
                return;
            }

            Node current = front;
            Console.Write("Front -> ");
            while (current != null)
            {
                Console.Write("[" + current.Data + "] -> ");
                current = current.Next;
            }
            Console.WriteLine("NULL");
        }


    }

    // Queue usage
    class Program
    {
        static void Main(string[] args)
        {
            Queue q = new Queue();

            q.Enqueue(10);
            q.Enqueue(20);
            q.Enqueue(30);
            q.PrintQueue();   // Front -> [10] -> [20] -> [30] -> NULL

            Console.WriteLine("Front element: " + q.Peek()); // 10
            Console.WriteLine("Dequeued: " + q.Dequeue());   // 10
            q.PrintQueue();   // Front -> [20] -> [30] -> NULL

            q.Enqueue(40);
            q.PrintQueue();   // Front -> [20] -> [30] -> [40] -> NULL
        }
    }
}

// Output:
// Front -> [10] -> [20] -> [30] -> NULL
// Front element: 10
// Dequeued: 10
// Front -> [20] -> [30] -> NULL
// Front -> [20] -> [30] -> [40] -> NULL